### DISCUSSION TITLE
Return custom error message from hook

### DISCUSSION THREAD

**Discussion by muety** [pocketbase/pocketbase#3376] [Q&A] [answered]
This had probably been asked before, but couldn't find it anywhere. If I create an `onBeforeCreateRequest()` hook in JavaScript to perform some validation on the incoming object that is too complex and lengthy to be modeled as an API rule, is there a way to return a custom error message alongside the `400` response? Also, is this even the proper way to implement "data-level" validation?

**Comment by ganigeorgiev** [comment, marked as answer]
You can check https://pocketbase.io/docs/js-routing/#error-response. For example:

```js
onRecordBeforeCreateRequest((e) => {
    // ...

    throw new BadRequestError("Custom error message...", {
        // optional data
        "someField": new ValidationError("invalid_someField_value", "Invalid or missing someField value."),
    })
}, "yourCollection")
```

>  Also, is this even the proper way to implement "data-level" validation?

Yes, this could be one way to do it. For simpler checks the API rules can also be used (_but you don't have control over the error message in this case_), for example:
```js
@request.data.someField ~ "something" && ...
```
Note that the embedded JS engine comes with some limitations and if you are trying to use an external validation library it may not always work (_see also [JSVM - Caveats and limitations](https://pocketbase.io/docs/js-overview/#caveats-and-limitations)_).

**Reply by muety** [reply]
Awesome, thanks. Pocketbase is such a f*cking cool project! :raised_hands: :pray:

**Comment by muety** [comment]
Just started a new project with PocketBase and came across the problem that custom error messages are not being passed on in the API response anymore. When throwing a `BadRequestError` from within my JavaScript hooks, the response will still only include a generic "Failed to create record.".

**Reply by ganigeorgiev** [reply]
Please open a new Q&A discussion and provide a code sample illustrating that.

With PocketBase v0.23+ API errors are allowed to be thrown from any hook (not just request hooks).

_edit: discussion continued at https://github.com/pocketbase/pocketbase/discussions/6102._
